Intersect.Client/Entities/Critter.cs

```diff
@@ -75,78 +75,98 @@ public override bool Update()
+                    break;
+                case Direction.UpRight:
+                    deltaX = 1;
+                    deltaY = -1;
+                    break;
+                case Direction.DownLeft:
+                    deltaX = -1;
+                    deltaY = 1;
+                    break;
+                case Direction.DownRight:
+                    deltaX = 1;
+                    deltaY = 1;
+                    break;
+            }
 
-                        break;
-                    case 3: // Right
-                        if (IsTileBlocked(X + 1, Y, Z, MapId, ref blockedBy, true, true, mAttribute.IgnoreNpcAvoids) == -1 && X < Options.MapWidth - 1 && (!mAttribute.BlockPlayers || !PlayerOnTile(MapId, X + 1, Y)))
-                        {
-                            //If BlockPlayers then make sure there is no player here
-                            tmpX++;
-                            IsMoving = true;
-                            Dir = 3;
-                            OffsetY = 0;
-                            OffsetX = -Options.TileWidth;
-                        }
+            var isBlocked = IsTileBlocked(tmpX + deltaX, tmpY + deltaY, Z, MapId, ref blockedBy, true, true,
+                mAttribute.IgnoreNpcAvoids) == -1;
+            var playerOnTile = PlayerOnTile(MapId, tmpX + deltaX, tmpY + deltaY);
 
-                        break;
-                }
+            if (isBlocked && tmpX + deltaX >= 0 && tmpX + deltaX < Options.MapWidth && tmpY + deltaY >= 0 &&
+                tmpY + deltaY < Options.MapHeight && (!mAttribute.BlockPlayers || !playerOnTile))
+            {
+                tmpX += (sbyte)deltaX;
+                tmpY += (sbyte)deltaY;
+                IsMoving = true;
+                Dir = MoveDir;
 
-                if (IsMoving)
+                if (deltaX == 0)
                 {
-                    X = (byte)tmpX;
-                    Y = (byte)tmpY;
+                    OffsetX = 0;
+                }
+                else
+                {
+                    OffsetX = deltaX > 0 ? -Options.TileWidth : Options.TileWidth;
+                }
 
-                    //TryToChangeDimension();
-                    MoveTimer = Timing.Global.Milliseconds + (long)GetMovementTime();
+                if (deltaY == 0)
+                {
+                    OffsetY = 0;
                 }
                 else
                 {
-                    if (MoveDir != Dir)
-                    {
-                        Dir = (byte)MoveDir;
-                    }
+                    OffsetY = deltaY > 0 ? -Options.TileHeight : Options.TileHeight;
                 }
             }
+            else if (MoveDir != Dir)
+            {
+                Dir = MoveDir;
+            }
+
+            if (!IsMoving)
+            {
+                return;
+            }
+
+            X = (byte)tmpX;
+            Y = (byte)tmpY;
+            MoveTimer = Timing.Global.Milliseconds + (long)GetMovementTime();
```

Comment:
I am pretty sure this is incorrect, and you should restore the `if (IsMoving) ...` and below to what it was before this change (excluding the (byte)MoveDir, that still needs to change).

The original code _set_ IsMoving inside of the if statement that you are "else if"-ing here.

Reply:
```csharp
            if (IsMoving)
            {
                X = (byte)tmpX;
                Y = (byte)tmpY;
                MoveTimer = Timing.Global.Milliseconds + (long)GetMovementTime();
            }
            else if (MoveDir != Dir)
            {
                Dir = MoveDir;
            }
```